cpp/src/arrow/compute/kernels/comparator.cc

```diff
@@ -0,0 +1,200 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/compute/kernels/comparator.h"
+
+#include "arrow/compute/context.h"
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/kernels/filter.h"
+#include "arrow/compute/kernels/util-internal.h"
+#include "arrow/util/bit-util.h"
+#include "arrow/util/logging.h"
+
+namespace arrow {
+
+namespace compute {
+
+class FunctionContext;
+struct Datum;
+
+static inline Status CopyOrTransferBitmap(const Array& array, MemoryPool* pool,
+                                          std::shared_ptr<Buffer>* bitmap) {
+  auto offset = array.offset();
+  auto length = array.length();
+  auto null_bitmap = array.null_bitmap();
+
+  if (offset == 0) {
+    // No offset adjustment to be done, take a zero-copy reference to the buffer.
+    *bitmap = null_bitmap;
+    return Status::OK();
+  } else if (null_bitmap == nullptr || length == 0) {
+    *bitmap = nullptr;
+    return Status::OK();
+  }
+
+  return internal::CopyBitmap(pool, null_bitmap->data(), offset, length, bitmap);
+}
+
+template <typename ArrowType, CompareOperator Op,
+          typename ArrayType = typename TypeTraits<ArrowType>::ArrayType,
+          typename ScalarType = typename TypeTraits<ArrowType>::ScalarType,
+          typename T = typename TypeTraits<ArrowType>::CType>
+static Status CompareArrayScalar(const ArrayType& input, const ScalarType& scalar,
+                                 uint8_t* bitmap) {
+  const T right = scalar.value;
+  const T* values = input.raw_values();
+
+  // Bitmap is not sliced like the original array, thus offset=0
+  auto writer = internal::BitmapWriter(bitmap, 0, input.length());
+  for (int64_t i = 0; i < input.length(); i++, writer.Next()) {
+    if (Comparator<T, Op>::Compare(values[i], right)) {
+      writer.Set();
+    } else {
+      writer.Clear();
+    }
+  }
+
+  writer.Finish();
+
+  return Status::OK();
+}
+
+template <typename ArrowType, CompareOperator Op>
+class CompareFunction final : public FilterFunction {
+  using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
+  using ScalarType = typename TypeTraits<ArrowType>::ScalarType;
+
+ public:
+  explicit CompareFunction(FunctionContext* ctx) : ctx_(ctx) {}
+
+  Status Filter(const Array& input, const Scalar& scalar, ArrayData* output) const {
+    // Caller must cast
+    DCHECK(input.type()->Equals(scalar.type));
+    // Caller must not pass null scalar, use IsNull, IsNotNull
```

Comment:
will the caller always have this option?  It seems reasonable to return false on all comparisons if null is passed in?

Reply:
We can't return false

```
sqlite> .nullvalue NULL
sqlite> SELECT 1 > NULL, NULL == NULL, 1 > 2;
NULL|NULL|0
```

Reply:
Ahh, sorry, I think I was thinking of where clauses

Reply:
It happens that the WHERE clause coalesce NULL to false. I intented to add an option to comparators where it would "merge" the valid_bitmap and result bitmap if the caller knows it'll coalesce.

This will be relevant in the next iteration with vectorization since it can be fused in the hot-loop (as opposed to use a sub-sequent AndOp).

Reply:
Updated to support this case